File: backend-node/routes/public.js
// ============================================================
// routes/public.js — Public read-only endpoints (no auth)
// Consumed by the main storefront (index.html)
// ============================================================

const express    = require('express');
const mongoose   = require('mongoose');
const Banner     = require('../models/Banner');
const CmsSection = require('../models/CmsSection');

const router = express.Router();

// GET /api/banners?type=hero|promotional|sidebar|festival
// Returns active banners visible today, sorted by displayOrder
router.get('/banners', async (req, res, next) => {
  try {
    if (mongoose.connection.readyState !== 1) {
      return res.json({ success: true, data: [] });
    }

    const { type } = req.query;
    const now = new Date();

    const filter = {
      status: { $in: ['active', 'approved'] },
      $or: [
        { endDate: null },
        { endDate: { $exists: false } },
        { endDate: { $gte: now } },
      ],
    };
    if (type) filter.type = type;

    const banners = await Banner
      .find(filter)
      .sort({ displayOrder: 1, createdAt: -1 })
      .select('title subtitle buttonText buttonUrl imageDesktop imageMobile type displayOrder');

    res.json({ success: true, data: banners });
  } catch (err) { next(err); }
});

// GET /api/cms/:key
// Returns a single CMS section by key if visible
router.get('/cms/:key', async (req, res, next) => {
  try {
    if (mongoose.connection.readyState !== 1) {
      return res.json({ success: true, data: null });
    }

    const section = await CmsSection.findOne({
      sectionKey: req.params.key,
      isVisible:  true,
    }).select('-lastEditedBy');

    res.json({ success: true, data: section || null });
  } catch (err) { next(err); }
});

module.exports = router;

File: backend-python/routes/public.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter

from config.database import get_db, is_connected
from utils.helpers import serialize_doc, serialize_list

router = APIRouter()


# GET /api/banners?type=hero|promotional|sidebar|festival
@router.get("/banners")
async def get_banners(type: Optional[str] = None):
    if not is_connected():
        return {"success": True, "data": []}

    db = get_db()
    now = datetime.utcnow()
    filt: dict = {
        "status": {"$in": ["active", "approved"]},
        "$or": [
            {"endDate": None},
            {"endDate": {"$exists": False}},
            {"endDate": {"$gte": now}},
        ],
    }
    if type:
        filt["type"] = type

    cursor = db.banners.find(filt).sort([("displayOrder", 1), ("createdAt", -1)])
    banners = serialize_list(await cursor.to_list(length=50))
    return {"success": True, "data": banners}


# GET /api/cms/:key
@router.get("/cms/{key}")
async def get_cms_section(key: str):
    if not is_connected():
        return {"success": True, "data": None}

    db = get_db()
    section = await db.cmssections.find_one({"sectionKey": key, "isVisible": True}, {"lastEditedBy": 0})
    return {"success": True, "data": serialize_doc(section)}
